### PR TITLE
Changed group structure for cohomology + new proof of ∥ Sⁿ ∥ₙ ≃ Ω ∥ Sⁿ⁺¹ ∥ₙ₊₁

### PR DESCRIPTION
**Pull request by aljungstrom**
This PR contains:

1. A new addition on the Eilenberg MacLane spaces ∥ Sⁿ ∥ₙ → ∥ Sⁿ ∥ₙ →∥ Sⁿ ∥ₙ  which is directly defined for n = 1 and sort-of-directly defined using the wedge connectivitiy lemma for spheres when n ≥ 2. The addition provably agrees with the previous one (by `genAddId` in `ZCohomology.GroupStructure`). This addition has two neat properties:
- 0ₖ + 0ₖ = 0ₖ - 0ₖ = 0ₖ now holds definitionally
- All group laws for Kₙ at the base point of Kₙ (n ≥ 1) are equal (sometimes definitionally) to refl
These two properties make path algebra a lot easier.

2. A direct proof of the fact that ΩKn is commutative w.r.t. path composition. This uses the new definition of + and should speed up some computations significantly.

3. New proofs in `ZCohomology.Groups`. In particular Hⁿ(A ⋁ B) ≅ Hⁿ(A) × Hⁿ(B) and Hⁿ(Sⁿ) ≅ ℤ have been given more synthetic proofs that don't rely on Mayer-Vietoris. In fact, Mayer-Vietoris is not used anywhere now. I added some explanations of some of the proofs. Maybe I overdid it, so tell me if I should remove something.

4. Some stengthened induction principles for cohomology groups (spread out in `ZCohomology.Groups`). Pretty much all of the old `isHom` proofs broke because of the new definition of addition, so I got lazy and gave these induction principles to help me prove everything again. They are probably completely useless for computations, but they are not really used anywhere anyway, so this shouldn't matter.

5. Some other tiny lemmas here and there. In particular, I've tried to make things used in `ZCohomology` more direct (e.g. the fact that x ≡ y is connected for x y : Sⁿ, and similar things) to help agda a bit.

I have also moved the old ZCohomology folder into experiments. This is now `Experiments.ZCohomologyOld`. So no need to check this part of the PR -- it's an exact copy of what is in the library now.

The cohomology groups of T² and S² ∨ S¹ ∨ S¹ compute a lot better than before for n = 1. Addition is pretty fast. For n = 2, there is some impovement. The map ℤ → H²(T²) → ℤ is pretty fast. Addition in H²(T²) works for some very trivial cases, but is incredibly slow. For H²(S² ∨ S¹ ∨ S¹), things are still stuck. It's a bit odd since the proofs here were a lot easier to carry out, but for some reason agda doesn't like them very much.

Sorry for this massive PR (however, much of it is just Experiments.ZCohomologyOld being introduced). I wanted to make sure that the new addition actually worked before I pushed anything.

Update:
Added some more stuff:
- An encode-decode proof of ∥ Sⁿ ∥ₙ ≃ Ω ∥ Sⁿ⁺¹ ∥ₙ₊₁ (using the new addition on ∥ Sⁿ ∥ₙ)
- Some basic facts about morphisms in `Cubical.Functions.Morphism`. I use this to avoid messy code in `ZCohomology` (since it's already so messy).

### CODE REVIEW

File: Cubical/ZCohomology/Properties.agda [hunk 1]
```diff
@@ -152,329 +173,201 @@ coHomRed+1Equiv zero A i = ∥ helpLemma {C = (Int , pos 0)} i ∥₂
...
 
-_+ₖ_ : {n : ℕ} → coHomK n → coHomK n → coHomK n
-_+ₖ_ {n = n} x y  = ΩKn+1→Kn n (Kn→ΩKn+1 n x ∙ Kn→ΩKn+1 n y)
+------------------- Kₙ ≃ ΩKₙ₊₁ ---------------------
 
--ₖ_ : {n : ℕ} →  coHomK n → coHomK n
--ₖ_ {n = n} x = ΩKn+1→Kn n (sym (Kn→ΩKn+1 n x))
+-- This proof uses the encode-decode method rather than Freudenthal
+-- The Code will be defined using the following Iso
+IsoKₙKₙ : (n : ℕ) (x : coHomK (suc n)) → Iso (coHomK (suc n)) (coHomK (suc n))
```
Comment (aljungstrom) [inline Cubical/ZCohomology/Properties.agda:181]: isn't this just addIso?

**Review by aljungstrom — COMMENTED** [review]
Check imports

File: Cubical/ZCohomology/GroupStructure.agda [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+                     × ((x : S₊ (2 + n)) → f x north ≡ ∣ x ∣)
+preAdd n =
+  wedgeConSn _ _
+             (λ _ _ → wedgeConHLev n)
+             ∣_∣
+             ∣_∣
+             refl
+
+-- addition for n = 1
+wedgeMapS¹ : S¹ → S¹ → S¹
```
[the remark this answers is not in the document]
Comment (aljungstrom) [reply Cubical/ZCohomology/GroupStructure.agda:101]: no Axel, this is slightly different. It makes cong _+_ loop loop reduce to loop*loop, which is probably why you picked it.

[the remark this answers is not in the document]
Comment (aljungstrom) [reply Cubical/ZCohomology/GroupStructure.agda:101]: thanks Axel, this makes sense.

File: Cubical/ZCohomology/GroupStructure.agda [hunk 3]
```diff
@@ -0,0 +1,630 @@
...
+  helper = sphereElim (suc n) (λ _ → isOfHLevelTrunc (4 + n) _ _ _ _)
+                              (sym (lUnit (sym (rUnitₖ (2 + n) (∣ north ∣ +ₖ ∣ north ∣)))))
+
+lUnitₖ≡rUnitₖ : (n : ℕ) → lUnitₖ n (coHom-pt n) ≡ rUnitₖ n (coHom-pt n)
+lUnitₖ≡rUnitₖ zero = isSetInt _ _ _ _
+lUnitₖ≡rUnitₖ (suc zero) = refl
+lUnitₖ≡rUnitₖ (suc (suc n)) = refl
+
+-- ΩKₙ is commutative w.r.t. path composition
+isComm∙ : ∀ {ℓ} (A : Pointed ℓ) → Type ℓ
```
[the remark this answers is not in the document]
Comment (aljungstrom) [reply Cubical/ZCohomology/GroupStructure.agda:247]: Cyclic dependencies. Moving to Homotopy.Loopspace instead

File: Cubical/ZCohomology/GroupStructure.agda [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+isComm∙ A = (p q : typ (Ω A)) → p ∙ q ≡ q ∙ p
+
+abstract
+  isCommA→isCommTrunc : ∀ {ℓ} {A : Pointed ℓ} (n : ℕ) → isComm∙ A
```
[the remark this answers is not in the document]
Comment (aljungstrom) [reply Cubical/ZCohomology/GroupStructure.agda:251]: Moved to Homotopy.Loopspace

**Comment by mortberg** [comment]
Great stuff!

**Pull request merged**